repose-aggregator/components/filters/header-normalization/src/main/scala/org/openrepose/filters/headernormalization/HeaderNormalizationFilter.scala

```diff
@@ -0,0 +1,133 @@
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
+import org.openrepose.core.filter.FilterConfigHelper
+import org.openrepose.core.filters.HeaderNormalization
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.core.services.reporting.metrics.MetricsService
+import org.openrepose.core.services.reporting.metrics.impl.MeterByCategorySum
+import org.openrepose.filters.headernormalization.config.{HttpMethod, HeaderNormalizationConfig}
+
+import scala.collection.JavaConverters._
+import scala.util.matching.Regex
+
+@Named
+class HeaderNormalizationFilter @Inject()(configurationService: ConfigurationService, metricsService: MetricsService)
+  extends Filter with UpdateListener[HeaderNormalizationConfig] with LazyLogging {
+
+  import HeaderNormalizationFilter._
+
+  private var configurationFile: String = DEFAULT_CONFIG
+  private var initialized = false
+  private var config: Seq[Target] = _
+  private var metricsMeter: Option[MeterByCategorySum] = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    logger.trace("Header Normalization filter initializing...")
+    configurationFile = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+
+    logger.info(s"Initializing Header Normalization filter using config $configurationFile")
+    val xsdUrl: URL = getClass.getResource(SCHEMA_FILE_NAME)
+    configurationService.subscribeTo(filterConfig.getFilterName, configurationFile, xsdUrl, this, classOf[HeaderNormalizationConfig])
+
+    metricsMeter = Option(metricsService) match {
+      case Some(service) =>
+        logger.debug("Initializing metrics meter for Header Normalization filter")
+        Some(service.newMeterByCategorySum(classOf[HeaderNormalization], "header-normalization", "Normalization", TimeUnit.SECONDS))
+      case None =>
+        logger.warn("Metrics meter not started for Header Normalization filter due to missing metrics service")
+        None
+    }
+
+    logger.trace("Header Normalization filter initialized.")
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    val wrappedRequest = new HttpServletRequestWrapper(servletRequest.asInstanceOf[HttpServletRequest])
+
+    config.find { target =>
+      // find the first "target" config element that matches this request (if any)
+      target.url.findFirstIn(wrappedRequest.getRequestURL).isDefined &&
+        (target.methods.contains(wrappedRequest.getMethod) || target.methods.contains(AllHttpMethods))
+    }.foreach { target =>
+      // figure out which headers to remove, and remove them
+      (target.access match {
+        case WhiteList => wrappedRequest.getHeaderNamesScala.diff(target.headers)
+          // written like this to maintain the case-insensitive string comparisons - do not swap
+        case BlackList => target.headers.intersect(wrappedRequest.getHeaderNamesScala)
```

Comment:
It's quite annoying.  That's why I put the note about not swapping the order because I know someone down the line will want to.